dapp/contracts/Crowdsale.sol

```diff
@@ -0,0 +1,210 @@
+    string public name = "GetLine Token";
+    string public symbol = "GET";
+    uint8 public decimals = 3;
+
+    /**
+     * GetToken contructor. Creates the GET token with a fixed supply all
+     * belonging to an owner.
+     * @param _treasury - The owner of all the initial supply of the token.
+     * @param _totalSupply - The initial supply of the token.
+     */
+    function GetToken(address _treasury, uint256 _totalSupply) public {
+        totalSupply = _totalSupply;
+        balances[_treasury] = _totalSupply;
+    }
+}
+
+contract BosonApproved is IBosonCallbackReceiver {
+    IBosonAddressResolver public bosonResolver;
+
+    event ValidationRequest(address _contract, address _who);
+    event ValidationResult(address _contract, address _who, bool _result);
+    struct Validations {
+        uint256 approvals;
+        uint256 rejections;
+    }
+    mapping(bytes32 => address) validationRequests;
+    mapping(address => Validations) validationResults;
+
+    function approved(address _address) public view returns (bool _approved) {
+        _approved = validationResults[_address].approvals > 0;
+    }
+
+    function rejected(address _address) public view returns (bool _rejected) {
+        _rejected = validationResults[_address].rejections > 0;
+    }
+
+    function approve(address _address) public {
+        if (approved(_address)) {
+            return;
+        }
+
+        ValidationRequest(this, _address);
+        bytes32 id = boson().queryAddress(_address);
+        validationRequests[id] = _address;
+    }
+
+    function __bosonCallback(bytes32 _id, bool _status) public {
+        require(msg.sender == boson().cbAddress());
+        address requested = validationRequests[_id];
+        delete validationRequests[_id];
+        ValidationResult(this, requested, _status);
+        if (_status) {
+            validationResults[requested].approvals++;
+            validationResults[requested].rejections = 0;
+        } else {
+            validationResults[requested].rejections++;
+        }
+    }
+
+    function boson() private returns (IBoson _boson) {
+        _boson = IBoson(bosonResolver.getAddress());
+    }
+
+}
+
+contract Crowdsale is Ownable, BosonApproved {
+    uint256 public GOAL_WEI = 10000 ether;
+    uint256 public DURATION = 14 days;
+    uint256 public MINIMUM_PURCHASE_WEI = 1 ether / 100;
+
+    using SafeMath for uint256;
+
+    enum State {
+        Initialized,
+        Started,
+        Finished
+    }
+
+    State public state;
+    GetToken public token;
+
+    uint256 public tokensLeft;
+    uint256 public tokensGoal;
+    uint256 public startDate;
+    uint256 public endDate;
+
+    event StateTransition(uint256 _oldState, uint256 _newState);
+    event StateTimeout(uint256 _oldState);
+
+    function nextState(State _newState) private {
+        StateTransition(uint256(state), uint256(_newState));
+        state = _newState;
+    }
+
+    function processTimeouts() private {
```

Comment:
Might want to allow to public so that we can manually change state after the timeout